Tahi-project/tahi
# APERTA-6321: Handling of salesforce submission errors

JIRA issue: [APERTA-6321](https://developer.plos.org/jira/browse/APERTA-6321)
#### What this PR does:

Sync'ing information with Salesforce happens in background jobs. In the past we relied on Sidekiq's built-in error handling and retry logic whenever we encountered an error.  This PR makes a separation between unknown and known errors as it relates to sync'ing information with Salesforce.

For the purpose of this PR:
- an _unknown error_ is something we didn't expected in the normal usage of the app. This are all still handled by Sidekiq's error handling and retry logic.
- a _known error_ is something we could anticipate. The example for this is if Aperta tries to sync a paper to salesforce that is missing a billing task or a financial disclosure task. 

In the _known error_ case above the PO (James) has requested that we log an error, email journal admins, and then stop retrying to sync that paper to salesforce since it will require human intervention to resolve the missing information.
#### Notes

The biggest surprise is that I reorganized some of the Salesforce sync'ing code while trying to keep my grubby little paws off the existing salesforce translation work. 

The code now flows like this:

```
SalesforceManuscriptUpdateWorker 
  -> SalesforceServices.sync_paper
     -> SalesforceServices::PaperSync
        ->  SalesforceServices::API.find_or_update_manuscript
     -> SalesforceServices::BillingSync
        -> SalesforceServices::API.ensure_pfa_case 
```

Here are my notes of why this pieces exist:
- `SalesforceManuscriptUpdateWorker` is responsible for kicking off a salesforce sync and handling errors. For known errors rescue them appropriately. For unknown error conditions pass thru to Sidekiq's error handling and retry logic. Thanks to @ajhekman for the sidekiq tips on error handling.
- `SalesforceServices.sync_paper!(...)` is responsible for handling at a high level what it means to sync a paper and its billing information to salesforce.
- `SalesforceServices::PaperSync` is responsible for validating the details of a paper before
  we send it to Salesforce. Uses `ActiveModel::Validations`.
- `SalesforceServices::BillingSync` is responsible for validating the details of a paper's billing information before sending it to Salesforce. Uses `ActiveModel::Validations`.
- The `SalesforceServices::ObjectTranslations` has been kept intact. I did remove the need to look up answers to questions on tasks since they can be looked up with one less dependency hop by going straight thru the paper. At some point these translations may be able to be merged with the sync objects, but it wasn't needed at this time so I didn't tackle that.
#### Major UI changes

n/a
#### How to test

**Step 1:** Get the salesforce test login credentials off from staging:

```
heroku config --app tahi-staging | grep DATABASEDOTCOM
```

Put those in your `.env`, `.env.development`, or `.env.local`.

**Step 2:** Log in to test.salesforce.com

Using the host, username, and password from the DATABASEDOTCOM credentials above go to http://test.salesforce.com and log in. 

**Step 3:** Testing the data

Reset your database using `rake db:reset` and fire up `rails console`. We'll test a few different cases:

**Test case 1:** Billing card is present, but some information is missing

The below snippet should be run in `rails console` and it should be successful in creating a Manuscript in Salesforce, but fail at creating a Case. This works because the seed data currently does not have all of the billing information filled out for Paper 3.

```
paper = Paper.find(3) 
paper. salesforce_manuscript_id = nil
paper.doi = "10.1371/MAKE-SOMETHING-UP-HERE-1"
paper.save!

PlosBilling::SalesforceManuscriptUpdateWorker.new.perform(paper.id)
```

If you go to your browser where you are logged into Salesforce you can click on the `Manuscripts` tab (or refresh if you're already there. You should see the Manuscript you just made at the top (or near the top). If you click into it and scroll down to the `Case` section it should have no cases listed. In fact, you should have seen an `Databasedotcom::SalesForceError` error printed.

**Test case 2:** Billing card is present and all information exists
- Run Aperta locally
- Log in as an admin
- Go to the workflow pages for Paper 3, and toggle the paper's editable flag on
- Go to the Billing Information card on the paper and click "Make changes to this card"
- Fill in all of the PFA information
- Go back to `rails console` and run the following

```
paper = Paper.find(3) 
paper. salesforce_manuscript_id = nil
paper.doi = "10.1371/MAKE-SOMETHING-ELSE-UP-HERE-1"
paper.save!

PlosBilling::SalesforceManuscriptUpdateWorker.new.perform(paper.id)
```

Go back to salesforce in the browser and you should another manuscript and this time when you try to find its case it will be there! :)

**Test case 3:** Financial Disclosure Card/Task does not exist

In `rails console` delete the financial disclosure card on the paper and sync to Salesforce:

```
paper = Paper.find(3) 
paper.financial_disclosure_task.destroy
paper. salesforce_manuscript_id = nil
paper.doi = "10.1371/MAKE-SOMETHING-ELSE-UP-HERE-1"
paper.save!

PlosBilling::SalesforceManuscriptUpdateWorker.new.perform(paper.id)
```

This should blow up (but in a safe way). You should get a HUGE backtrace and if you scroll up you should get some human readable information as it relates to the error that was thrown. If you look at your development log you should see this is emailed out to the journal admins.

**Test case 4:** Billing Card/Task does not exist

You may want to `rake db:reset` before doing this since we want to isolate the fact that the billing card does not exist. 

Then, in `rails console` delete the billing card on the paper and sync to Salesforce:

```
paper = Paper.find(3) 
paper.billing_task.destroy
paper. salesforce_manuscript_id = nil
paper.doi = "10.1371/MAKE-SOMETHING-ELSE-UP-HERE-1"
paper.save!

PlosBilling::SalesforceManuscriptUpdateWorker.new.perform(paper.id)
```

This should also blow up (but in a safe way). You should once again get a HUGE backtrace and if you scroll up you should get some human readable information as it relates to the error that was thrown. If you look at your development log you should see this is emailed out to the journal admins.

---
#### Code Review Tasks:

Author tasks:  

Reviewer tasks:
#### After the Code Review:

Author tasks:
